s3/src/main/scala/blobstore/s3/S3Store.scala

```diff
@@ -477,14 +462,98 @@ object S3Store {
   private val maxMultipartParts: Int           = 10000
 
   private val multipartUploadPartsError = new IllegalArgumentException(
-    s"S3 doesn't support multipart uploads with more than ${S3Store.maxMultipartParts} parts."
+    show"S3 doesn't support multipart uploads with more than ${S3Store.maxMultipartParts} parts."
   )
 
-  private val multipartUploadBufferTooSmallError = new IllegalArgumentException(
-    s"Please use buffer size of at least 5Mb – S3 requires minimal size of 5Mb per part of multipart upload."
-  )
+  /** @see
+    *   [[S3Store]]
+    */
+  trait S3StoreBuilder[F[_]] {
+    def withS3Client(s3Client: S3AsyncClient): S3StoreBuilder[F]
+    def setCrtClient(maybeCrtClient: Option[S3AsyncClient]): S3StoreBuilder[F]
+    def setObjectAcl(maybeObjectAcl: Option[ObjectCannedACL]): S3StoreBuilder[F]
+    def setSseAlgorithm(maybeSseAlgorithm: Option[ServerSideEncryption]): S3StoreBuilder[F]
+    def withBufferSize(bufferSize: Int): S3StoreBuilder[F]
+    def withQueueSize(queueSize: Int): S3StoreBuilder[F]
+    def enableFullMetadata: S3StoreBuilder[F]
+    def disableFullMetadata: S3StoreBuilder[F]
+    def enableTrailingSlashFiles: S3StoreBuilder[F]
+    def disableTrailingSlashFiles: S3StoreBuilder[F]
+    def withCrtClient(crtClient: S3AsyncClient): S3StoreBuilder[F]              = setCrtClient(Some(crtClient))
+    def withObjectAcl(objectAcl: ObjectCannedACL): S3StoreBuilder[F]            = setObjectAcl(Some(objectAcl))
+    def withSseAlgorithm(sseAlgorithm: ServerSideEncryption): S3StoreBuilder[F] = setSseAlgorithm(Some(sseAlgorithm))
+    def build(): ValidatedNec[Throwable, S3Store[F]]
+    def unsafe(): S3Store[F] = build() match {
+      case Validated.Valid(a)    => a
+      case Validated.Invalid(es) => throw es.reduce(Throwables.collapsingSemigroup) // scalafix:ok
+    }
+  }
 
-  private val multipartUploadBufferTooLargeError = new IllegalArgumentException(
-    s"Please use buffer size less than 500Mb – S3 requires maximum object size of 5Tb, and maximum number of 10000 parts in multipart upload."
-  )
+  case class S3StoreBuilderImpl[F[_]: Async](
+    _s3Client: S3AsyncClient,
+    _crtClient: Option[S3AsyncClient] = None,
+    _objectAcl: Option[ObjectCannedACL] = None,
+    _sseAlgorithm: Option[ServerSideEncryption] = None,
+    _defaultFullMetadata: Boolean = false,
+    _defaultTrailingSlashFiles: Boolean = false,
+    _bufferSize: Int = S3Store.multiUploadDefaultPartSize.toInt,
+    _queueSize: Int = 32
+  ) extends S3StoreBuilder[F] {
+    def withS3Client(s3Client: S3AsyncClient): S3StoreBuilder[F]               = this.copy(_s3Client = s3Client)
+    def setCrtClient(maybeCrtClient: Option[S3AsyncClient]): S3StoreBuilder[F] = this.copy(_crtClient = maybeCrtClient)
```

Comment:
Can we take `S3AsyncClient with S3CrtAsyncClient` and avoid that runtime type check?

Reply:
The reason I'm not referencing `S3CrtAsyncClient` by type and doing runtime reflection instead is that I didn't want to require preview dependency (`software.amazon.awssdk:s3-transfer-manager`) that transitively brings in CRT (10MB Jar on it's own) for an optional feature.